### DISCUSSION TITLE
M1 Monterey (betas) fork

### DISCUSSION THREAD

**Discussion by donaldguy** [asmvik/yabai#1042] [Show and tell]
I suspect everyone who cares knows already from comments in #923 - but **FYI**, I am/have been operating a fork (building on the M1 work of @xorpse et al) that you can find at:

https://github.com/donaldguy/yabai  

### Administrivia

- Detailed build/Install instructions can be found below: https://github.com/koekeishiya/yabai/discussions/1042#discussioncomment-1590929 
- I am a partisan of the fic-style  branch naming conventions proposed (probably jokingly) in https://twitter.com/tesseralis/status/1271197776886370305 ergo the trunk branch is [`canon`](https://github.com/donaldguy/yabai/tree/canon)

## The Present 

it (mostly) works on **12.1 beta [~~1 (21C5021h)~~ 2](https://github.com/donaldguy/yabai/commit/866869158f853995c5c42c78f1e40b46e66dc700) (21C5031d), beta 3 (21C5039b), and beta 4 (21C5045a)**  (which was binary compatible with beta 2 as far as Dock went) and **12.0.1 ~~RC 2~~[^1] (21A559)**

I am only running & checking on M1. There is a second `HEAD` branch (appropriately) named [`intel-maybe-probably-not`](https://github.com/donaldguy/yabai/tree/intel-maybe-probably-not), YMMV.

I have upgraded to 12.1 beta(s) and `canon` will continue to track those as they are released and I get them. 

If 12.0.2… is forthcoming @mak1A4 has [asserted](https://github.com/koekeishiya/yabai/issues/923#issuecomment-954235081) he will update (M1) offsets.

For either the 12.0.* or the intel case, if someone wants to tackle double checking / updating the SA offsets, I outlined my (fairly naive) binary diffing approach [here](https://github.com/koekeishiya/yabai/issues/923#issuecomment-899116832)

## The Future

@koekeishiya has an M1 pro on the way[^2] and will be targeting 12.0.1 on it when he can[^3]– so my fork is intended to be temporary and may become archived.

It may also get repurposed to run ahead of [`the-future`](https://github.com/koekeishiya/yabai/tree/the-future) for developer betas, which I am in the deeply-questionable habit of installing shortly after their release. 

That said, per request, Issues are open on the repo now and I will do my best to review PRs. 

Also, feel free (unless/until @koekeishiya says otherwise) to also talk in here.

[^1]: I believe which was promoted to final 12.0.1? (at least the Dock.app was the same?) 
[^2]: https://github.com/koekeishiya/yabai/issues/725#issuecomment-949484389 
[^3]: https://github.com/koekeishiya/yabai/issues/923#issuecomment-956198083

**Comment by jclusso** [comment]
@donaldguy appreciate your efforts. Could you add the proper procedure for using your fork. I'm struggling to figure out how to specify to brew to install your version.

**Reply by donaldguy** [reply]
**TL;DR** Install the upstream[^1] distro from homebrew and overwrite the binary (`/opt/homebrew/Cellar/yabai/*/bin/yabai`, not the symlink to it that's actually in your `PATH`) 

---
<details>
<summary> <-- Click for full breakdown of steps and command </summary>

This was cribbed somewhat from  @shepherdjerred in https://github.com/koekeishiya/yabai/issues/923#issuecomment-934617044  but expanding/editorialized quite a bit also:


1. Install upstream homebrew formulae package(s)

    ```bash
    brew install yabai #skhd
    ```

2. clone the fork (e.g. 

    ```
    git clone git@github.com:donaldguy/yabai.git
    ```
    )
3. Follow instructions in [README-arm64.md](https://github.com/donaldguy/yabai/blob/canon/README-arm64.md) (unchanged from @xorpse's fork, non-redundant parts transcluded for convenience:
    >    Enter recovery mode and disable SIP using:
    >
    >```
    >$ csrutil disable --with kext --with dtrace --with basesystem
    >```
    >
    >Reboot and add boot arguments and reboot again:
    >
    >```
    >$ sudo nvram boot-args=-arm64e_preview_abi
    >```

    ). I'll add that the recovery mode procedure [changed](https://support.apple.com/en-us/HT204904) for the M1 macs vs older intel ones and is basically just hold down power button during boot (like press it down to start and keep holding until ... it should be clear). You probably/may have to also use the [GUI "Startup Security Utility"](https://support.apple.com/en-us/HT208198) to downgrade that checking, but I am not sure.

4. (Back in non-recovery mode, reopen terminal and go back to where you cloned the repo, such that your pwd is `…/yabai` and) build the binary
    ```
    make
    ```
5. Create a signing cert if[^2] you plan to use `launchd` (via `brew services`) and don't already have one installed. 
Excerpting from https://github.com/koekeishiya/yabai/wiki/Installing-yabai-(from-HEAD):
    >First, open `Keychain Access.app`. In its menu, navigate to `Keychain Access`, then `Certificate Assistance`, then click  `Create a Certificate...`. This will open the `Certificate Assistant`. Choose these options:
    >- Name: `yabai-cert`,
    >- Identity Type: `Self-Signed Root`
    >- Certificate Type: `Code Signing`
6. `make sign` (or somewhat equivalently `./scripts/codesign ./bin/yabai` — this will use your only installed codesigning cert, if you have exactly one but its not (necessarily) named `yabai-cert`)
7.  Replace the homebrew binary with the one just built

    ```
     mv ./bin/yabai "$(realpath "$(which yabai)")"
    ```
    The indirection[^3] of target is required cause the launchd units from homebrew refer to the absolute installed path (<samp>/opt/homebrew/Cellar/yabai/<var>$versionNumber</var>/bin/yabai</samp>) 

8. If you are (planning on) using the Scripting Additions (and why else are you on this fork?) make sure they are installed right

    ```
     sudo yabai --uninstall-sa
     sudo yabai --install-sa
    ```

     consider, per https://github.com/koekeishiya/yabai/wiki/Installing-yabai-(from-HEAD)#macos-big-sur---automatically-load-scripting-addition-on-startup , adding ability to load SA `NOPASSWD`-style to your `sudoers` file, then adding

    ```
     sudo yabai --load-sa
    ```

    early in your `.yabairc` (but you can also run it manually each time, before (re)starting `yabai`)

9. Enable the service(s)

    ```
    brew services start yabai
    ```

    ```
    brew services start skhd
    ```



[^1]: Personally, I installed @xorpse's formula (`brew install --HEAD xorpse/formulae/yabai`) but it seems like the core tap one is probably fine too

[^2]: The cert identity, which is obvs not [Gatekeeper](https://en.wikipedia.org/wiki/Gatekeeper_(macOS)) approved anyway, is primarily is used to maintain the granted accessibility permissions across upgrades, so if you are always gonna run `yabai` interactively you can just grant to the Terminal (e.g. iTerm.app) instead.

[^3]: [`realpath`](https://man7.org/linux/man-pages/man1/realpath.1.html) can be installed with `brew install coreutils`, or like VSCode's cli (`code`), you could substitute `python -c "import os,sys; print(os.path.realpath(sys.argv[1]))" `, or you could replace `"$(realpath "$(which yabai)")"` with `/opt/homebrew/Cellar/yabai/*/bin/yabai` - iff you only have one version installed

</details>

**Reply by joshyork** [reply]
Thanks for putting this together. I'm on an M1 and Monterey (12.1 beta) and can confirm yabai is fully working for me again! I thought I was already following these steps during previous unsuccessful attempts. I'm not sure if upgrading to the beta (12.1 vs 12.0.1) was the difference maker or perhaps I was fudging one of the steps previously. Either way huge thanks to everyone involved for getting this working!

**Reply by jclusso** [reply]
@donaldguy thanks for that. I'm wondering how this works. In the past and even when it was broken I was able to get it partially working without disabling SIP. After I set it up can I re-enable SIP? Without it, I can't use some apps because they require it. Prior to Monterey this worked fine and I was able to use most of Yabai without an issue.

**Reply by neilsabde** [reply]
@donaldguy I followed all the steps and yabai starts up properly but when I open a new window in chrome, it is completely transparent and then I need to force quit it. Any idea on why this happens?

**Reply by neilsabde** [reply]
I'm using xorpse's fork and I have replaced the binary with the one in `yabai/bin/yabai`

**Reply by AStaroverov** [reply]
Thank you, everything is working!

**Comment by donaldguy** [comment]
I kinda doubt I can summon via mention into a discussion unbidden, but I'll try:

 @pschwartz !  In re https://github.com/koekeishiya/yabai/issues/923#issuecomment-960178256:
> I am using skhd with Yabai. I have compiled the latest from source and tried donaldguy/yabai's latest release. I am running a MacBook Air with the M1 and didn't read issues before upgrading to Monterey.
> 
> Yabai is working and Skid is. But none of the space focus commands provide any return or work. I can move a windows between spaces, but doing a `yabai -m space --focus 1` fails to move my current focus. The --check-sa has a return one of 0.
> 
> Any thoughts?



the one release on my fork is not up to date, it was just a one off bone-throw for the person who asked for one for [Nix](https://nixos.org) reasons.

Try building yourself per instructions [above](https://github.com/koekeishiya/yabai/issues/923#discussioncomment-1590929) 

---

... I'm a little squeamish encouraging people to install a pre-built binary of what like ... already behaves more than a little like a rootkit, even though I know I have no intention of abusing it. 

[Trusting Trust](https://dl.acm.org/doi/pdf/10.1145/358198.358210), moral hazard and all that.

but if someone wants to PR a GitHub Actions workflow for build and release, I'd consider merging it.

**Reply by joshtwo** [reply]
I also can't use focus space. Everything else works fine, even creating new workspaces.

**Reply by joshtwo** [reply]
This is what the error log shows:
```
could not retrieve window details.
sh: line 0: [: =: unary operator expected
could not retrieve window details.
sh: line 0: [: =: unary operator expected
could not retrieve window details.
sh: line 0: [: =: unary operator expected
could not retrieve window details.
sh: line 0: [: =: unary operator expected
could not retrieve window details.
sh: line 0: [: =: unary operator expected
```

**Reply by joshtwo** [reply]
Nevermind, I was loading the main repo yabai instead of the fork. I didn't notice until now.

**Reply by donaldguy** [reply]
For the record those error messages suggests that the immediate problem was that you were interpreting with `sh` a file (likely `~/.yabairc`) designed for `bash` or `zsh` ... I think

Likely using `==`s where a stricter POSIX `test` (aka `[`) wanted to see `=`s 

or something like that. (`==` is only technically valid in `[[` tests, but I think bash tends to be permissive about it)

**Comment by ppanagiotis** [comment]
Did someone tried this fork with 12.1 and intel cpu? Thank you!

**Reply by DennisFeldbusch** [reply]
+1

**Reply by ai212983** [reply]
Yeah, just tried it with intel CPU. Not compiling. Not sure if this problem has something to do with CPU though.

```
❯ make
makefile:42: warning: overriding commands for target `src/osax/sa_payload.c'
makefile:34: warning: ignoring old commands for target `src/osax/sa_payload.c'
rm -rf ./bin
/Library/Developer/CommandLineTools/usr/bin/clang ./src/osax/loader.m -shared -O2 -mmacosx-version-min=10.13 -arch x86_64 -o ./src/osax/loader -isysroot "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk" -framework Foundation
In file included from ./src/osax/loader.m:2:
In file included from ./src/osax/x86_64/loader.m:1:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/Foundation.h:12:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSBundle.h:91:143: error: function does not return NSString
- (NSAttributedString *)localizedAttributedStringForKey:(NSString *)key value:(nullable NSString *)value table:(nullable NSString *)tableName NS_FORMAT_ARGUMENT(1) NS_REFINED_FOR_SWIFT API_AVAILABLE(macos(12.0), ios(15.0), watchos(8.0), tvos(15.0));
                                                         ~~~~~~~~~~~~~~                                                                       ^                  ~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:103:48: note: expanded from macro 'NS_FORMAT_ARGUMENT'
        #define NS_FORMAT_ARGUMENT(A) __attribute__ ((format_arg(A)))
                                                      ^          ~
1 error generated.
make: *** [src/osax/sa_loader.c] Error 1
```

**Reply by ai212983** [reply]
although it must be old xcode, [here](https://github.com/koekeishiya/yabai/issues/923#issuecomment-953399516) @donaldguy describes how to fix it.

**Reply by ai212983** [reply]
Works on my machine (tm), [here's](https://github.com/koekeishiya/yabai/issues/923#issuecomment-962077404) instructions

**Reply by LordSushiPhoenix** [reply]
Also seems to work on my intel mac, aside from the `yabai -m space --focus N` command, which gives no output.

**Reply by donaldguy** [reply]
For the record, `-m space --focus` does seem to work for me on M1

Again, no one has sussed out the Intel SA injection points - which you would need for this to work, I believe.

The instructions for how to do so (how I do so on M1 anyway) are here: https://github.com/koekeishiya/yabai/issues/923#issuecomment-899116832

PRs welcome

**Reply by donaldguy** [reply]
It's also possible, if the assembled versions of the functions are unchanged from Big Sur, that all someone needs to do is add the `== 12` condition to the existing branch in

https://github.com/donaldguy/yabai/blob/866869158f853995c5c42c78f1e40b46e66dc700/src/osax/x86_64/payload.m#L204

etc

**Comment by donaldguy** [comment]
Made the single change for 12.1 beta 2: https://github.com/donaldguy/yabai/commit/866869158f853995c5c42c78f1e40b46e66dc700

doesn't seem like much touched Dock

**Comment by joshtwo** [comment]
I often get crashes when the computer sleeps. Is anyone else experiencing this? I've got a [crash log](https://gist.github.com/joshtwo/77184633829d6155ea3fea7a34236a2c) in case that helps.

**Reply by donaldguy** [reply]
It's a little useless without any debug symbols honestly.

If you are set up with Feedback Assistant, Apple will _sometimes_ offer you (or at least me 😅 ) the chance to submit a bug report (which you don't actually need to do—I assume they probably don't want anyway - since they'd rather people not futz around like this—but initially say you're gonna), and then while the sysdiagnose stuff is running you can do a Quick Look on the `.ips` file (which you can reveal in Finder from the submit form, then press spacebar on) and see (and copy/paste) a crash report _with_ symbols loaded (presumably after the fact?)

I've never been super clear on how/if you can force this. But that would at least tell us where in SkyLight its going down

I am not having this issue (nor do I think most people are), so I would try futzing around with your config and see if you can narrow it down. My guess is its maybe borders related (that's the area that was generally buggy enough that I've just given up for the time being)

EDIT: but its almost certainly not gonna be an option ever on a stable build like 21A344 where the Feedback Assistant isn't even part of the install.

**Reply by donaldguy** [reply]
(but it could also be an M1 Max issue? I am on a lowly M1: Original Recipe)

**Reply by donaldguy** [reply]
Also though, that logs says you are on build 21A344 (12.0.0) which I do _not_ have offsets for (and may or may not have ever - as it seems like the betas branched from the MBP pro flashing builds early, going by the numbering)

Consider upgrading to 12.0.1 (21A559) which I believe (cause e.g. [Wikipedia](https://en.wikipedia.org/wiki/MacOS_Monterey#Release_history) says so) went out on the public release channel? and these offsets should work on https://github.com/donaldguy/yabai/commit/00bf5e46685df0cd3e961c155615ddecc75562ab - still there at HEAD too)

Or make bad choices like meeeee and run 12.1 betas

**Reply by joshtwo** [reply]
I'll upgrade to 12.0.1, thanks a lot!

**Comment by donaldguy** [comment]
I ran the analysis / binary diff and the offsets from beta 2 (21C5031d) seem identical in beta 3 (21C5039b); I've updated the OP and repo description to reflect compatibility thereof

**Reply by donaldguy** [reply]
Ditto on beta 4 (21C5045a)

**Comment by donaldguy** [comment]
Have users of the fork checked out koekeishiya's new 4.0 prerelease?

I haven't yet but I'm mostly inclined to switch over to it - maybe bring over the xorpse-fork style SA injection if it seems more reliable

I may archive my fork entirely; but I'm open to keeping it going if there's a reason.

Any thoughts, anybody?

**Comment by ghost** [comment]
Hi, my Mac version is 12. 1.  I follow your Youtube video and download yabai successfully! I'm deeply appreciated your effort. 

However, I encounter the problem that when I use `make` command. I cannot complie the `extract.sh` file with some invaild instruction error. Firstly, I think it might be the assembly syntax difference bewteen intel and AT&T. After STFW, I didn't find a way to use a different syntax or even figure out the principle. But finally, I solve it through switch my ARCH from x86_64 to arm64 by 

`$env /usr/bin/arch -arm64 /bin/zsh`

And then, I compiled all the files smoothly. As I find some people with same problem in youtube comment, I just put my solution here and hopefully it cuold help others.

**Comment by donaldguy** [comment]
I archived the repo since 4.0 has been released
